File: web/src/dtos/Product.ts
import Common from "../utility/Common";
import { CategoryRes } from "./Category";
import { PagedReq } from "./PagedReq";
import { ProductImageRes } from "./ProductImage";

export interface ProductRes {
  productId?: number;
  name?: string;
  code?: string;
  description?: string;
  position?: number;
  quantity?: number;
  rate?: number;
  categoryId?: number;
  unitId?: number;

  category?: CategoryRes;

  productImages?: ProductImageRes[];
}

export class ProductReqEdit {
  name?: string = "";
  code?: string = "";
  description?: string = "";
  position?: number = 10;
  quantity?: number = 0;
  rate?: number = 0;
  categoryId?: number = 0;
  unitId?: number = 0;

  constructor(categoryId?: number) {
    this.categoryId = categoryId;
  }
}

export enum StockStatus {
  InStock = 1,
  OutOfStock = 0,
  AllProducts = -1,
}

export class ProductReqSearch extends PagedReq {
  categoryId?: number;
  categoryCode?: string;
  stockStatus?: number;

  constructor(
    {
      pageNumber = 1,
      pageSize = Common.DEFAULT_PAGE_SIZE,
      orderBy,
      searchText = "",
    }: PagedReq,
    { categoryId = 0, categoryCode = "", stockStatus = StockStatus.AllProducts }
  ) {
    super({
      pageNumber: pageNumber,
      pageSize: pageSize,
      orderBy: orderBy,
      searchText: searchText,
    });
    this.categoryId = categoryId;
    this.categoryCode = categoryCode;
    this.stockStatus = stockStatus;
  }
}

File: data/Dtos/Product.cs
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using data.Entities;
using data.Utility.Paging;

namespace data.Dtos
{
    public class ProductRes
    {
        public int ProductId { get; set; }
        public string? Name { get; set; }
        public string? Code { get; set; }
        public string? Description { get; set; }
        public int Position { get; set; } = 10;
        public decimal Quantity { get; set; }
        public decimal Rate { get; set; }
        public int? UnitId { get; set; }
        public Unit? Unit { get; set; }
        public int CategoryId { get; set; }
        public Category? Category { get; set; }
        public IEnumerable<ProductImageRes>? ProductImages { get; set; }
    }

    public class ProductReqEdit
    {
        public string? Name { get; set; }
        public string? Code { get; set; }
        public string? Description { get; set; }
        public int Position { get; set; } = 10;
        public decimal Quantity { get; set; }
        public decimal Rate { get; set; }
        public int? UnitId { get; set; }
        public int CategoryId { get; set; }
    }

    public class ProductReqSearch : PagedRequestDto
    {
        public int CategoryId { get; set; }
        public string? CategoryCode { get; set; }
        public int StockStatus { get; set; } = -1;
    }

    public enum StockStatus
    {
        InStock = 1,
        OutOfStock = 0,
        AllProducts = -1
    }
}
